### DISCUSSION TITLE
Superset 2.x: `AttributeError: module 'flask.json' has no attribute 'JSONEncoder'`

### DISCUSSION THREAD

**Discussion by amotl** [apache/superset#27211] [General]
### Bug description

Hi there,

we are running integration tests with Apache Superset and CrateDB on behalf of our [cratedb-examples](https://github.com/crate/cratedb-examples) repository. Two days ago, the corresponding CI job started failing with the error message `AttributeError: module 'flask.json' has no attribute 'JSONEncoder'`. The error happens when invoking `superset db upgrade`, and only on Apache Superset 2.x.

- https://github.com/crate/cratedb-examples/issues/313

It baffles us that the error is apparently happening out of the blue, without any obvious things being changed. Everything worked well until two days ago.

Based on suggestions gathered on behalf of https://github.com/crate/cratedb-examples/issues/313, we tried to fiddle with a few dependencies already, trying different combinations of versions around Flask and friends, but haven't been able to discover the root cause.

With kind regards,
Andreas.

### How to reproduce the bug

Invoke `superset db upgrade` on the most recent Apache Superset 2.x, specifically [Apache Superset 2.1.3](https://pypi.org/project/apache-superset/2.1.3/).

### Superset version

2.1.3

### Python version

3.11

### Node version

Not applicable

### Browser

Not applicable

### Additional context

```python
Traceback (most recent call last):
  File "/path/to/bin/superset", line 5, in <module>
    from superset.cli.main import superset
  File "/path/to/lib/python3.11/site-packages/superset/__init__.py", line 21, in <module>
    from superset.app import create_app
  File "/path/to/lib/python3.11/site-packages/superset/app.py", line 23, in <module>
    from superset.initialization import SupersetAppInitializer
  File "/path/to/lib/python3.11/site-packages/superset/initialization/__init__.py", line 27, in <module>
    from flask_appbuilder import expose, IndexView
  File "/path/to/lib/python3.11/site-packages/flask_appbuilder/__init__.py", line 5, in <module>
    from .api import ModelRestApi  # noqa: F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/path/to/lib/python3.11/site-packages/flask_appbuilder/api/__init__.py", line 38, in <module>
    from ..baseviews import AbstractViewApi
  File "/path/to/lib/python3.11/site-packages/flask_appbuilder/baseviews.py", line 22, in <module>
    from .forms import GeneralModelConverter
  File "/path/to/lib/python3.11/site-packages/flask_appbuilder/forms.py", line 3, in <module>
    from flask_wtf import FlaskForm
  File "/path/to/lib/python3.11/site-packages/flask_wtf/__init__.py", line 4, in <module>
    from .recaptcha import Recaptcha
  File "/path/to/lib/python3.11/site-packages/flask_wtf/recaptcha/__init__.py", line 1, in <module>
    from .fields import RecaptchaField
  File "/path/to/lib/python3.11/site-packages/flask_wtf/recaptcha/fields.py", line 3, in <module>
    from . import widgets
  File "/path/to/lib/python3.11/site-packages/flask_wtf/recaptcha/widgets.py", line 6, in <module>
    JSONEncoder = json.JSONEncoder
                  ^^^^^^^^^^^^^^^^
AttributeError: module 'flask.json' has no attribute 'JSONEncoder'
```

### Checklist

- [ ] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [X] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [X] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by michael-s-molina** [comment]
Hi @amotl. I converted your issue to a GitHub discussion as 2.x versions are not actively supported anymore. I'll ping @dpgaspar who might have more context about Flask App Builder and it's dependencies.

Nevertheless, I highly recommend upgrading Superset to 3.1.

**Reply by amotl** [reply]
Hi @michael-s-molina. Thanks for your quick response. Our CI job is also continuously verifying Superset 3 compatibility with CrateDB, and that works very well, thanks!

- https://github.com/crate/cratedb-examples/pull/218

We already recognized Superset 2 is being out of service already. Nevertheless, we wanted to bring this to your attention in one way or another. Thanks for diverting it into a discussion.

Keep up the excellent work!

**Reply by amotl** [reply]
https://github.com/crate/cratedb-examples/pull/335 will remove testing of Apache Superset 2.x on our CI systems.

**Comment by Jagath01234** [comment]
I'm seeing this same error message with these versions. 
```
root@244c0f8b58e3:/app# superset --version
Loaded your LOCAL configuration at [/app/docker/pythonpath_dev/superset_config.py]
Python 3.10.14
Flask 2.3.3
Werkzeug 3.0.1
```
However the usecase is different. 
I'm seeing this when I'm triying to log in a user using  /api/v1/security/login.  the CURL is below if you want to reproduce it.

`curl --location --request POST 'http://localhost:8088/api/v1/security/login' \
--header 'accept: application/json' \
--header 'Content-Type: application/json' \
--data-raw '{
  "password": "admin",
  "provider": "db",
  "refresh": false,
  "username": "admin"
}'`
